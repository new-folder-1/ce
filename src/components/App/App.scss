body {
    font-family: 'Roboto', sans-serif;
    background-color: #2196f3;

    display: flex;
    justify-content: center;
}

.App {
    margin: 12px;
    box-sizing: border-box;

    &-Header {
        display: flex;
        justify-content: flex-end;
        padding: 12px 0;
    }

    &-Wallet {
        position: relative;
        padding: 24px;

        &_from {
            padding: 24px;
            background-color: #2c6cff;

            &::before {
                content: "";
                position: absolute;
                border: 8px solid transparent;
                border-top-color: #2c6cff;
                left: 50%;
                bottom: -16px;
                z-index: 1;
                transform: translateX(-50%);
            }
        }

        &_to {            
            background-color: #255dd3;
        }
    }
}
