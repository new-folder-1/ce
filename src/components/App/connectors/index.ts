import { Dispatch } from "redux";
import { fetchWalletsAsync, submitExchangeAsync, ExchangeSubmit } from "../../../store/wallets/actions";
import { connect } from "react-redux";
import { RootState } from "../../../store";
import { getWallets, getFetchError as walletsError, getExchangeError } from "../../../store/wallets/selectors";
import { updateBaseCurrency } from "../../../store/rates/actions";
import { Currency } from "../../../types";
import { getRates, getRateError as ratesError } from "../../../store/rates/selectors";

const mapStateToProps = (state: RootState) => {
    return {
        wallets: getWallets(state),
        rates: getRates(state),
        globalError: walletsError(state) || ratesError(state),
        exchangeError: getExchangeError(state)
    };
};

const mapDispatchToProps = (dispatch: Dispatch) => {
    return {
        fetchWallets: () => dispatch(fetchWalletsAsync.request()),
        updateBaseCurrency: (currency: Currency) => dispatch(updateBaseCurrency(currency)),
        submitExchange: (data: ExchangeSubmit) => dispatch(submitExchangeAsync.request(data))
    };
};

export const appConnector = connect(mapStateToProps, mapDispatchToProps);
