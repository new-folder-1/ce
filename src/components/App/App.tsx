import * as React from "react";

import { appConnector } from "./connectors";
import { Wallet, ExchangeRates, Currency } from "../../types";

import './App.scss';
import { WalletPicker, WalletType } from "../WalletPicker/WalletPicker";
import { ExchangeSubmit } from "../../store/wallets/actions";

interface AppProps {
    title: string;

    rates: ExchangeRates,
    wallets: Wallet[];
    fetchWallets: () => void;
    updateBaseCurrency: (currency: Currency) => void;
    submitExchange: (data: ExchangeSubmit) => void;
}

interface AppState {
    walletFrom?: Wallet;
    walletTo?: Wallet;

    amount: number;
}

class AppPresenter extends React.Component<AppProps, AppState> {
    constructor(props: AppProps) {
        super(props);

        this.state = {
            amount: 0
        };

        this.onBallanceChange = this.onBallanceChange.bind(this);
        this.onWalletChange = this.onWalletChange.bind(this);
        this.submitExchange = this.submitExchange.bind(this);
    }

    componentDidMount() {
        this.props.fetchWallets();
    }

    componentDidUpdate() {
        if (!this.state.walletFrom && !this.state.walletTo) {
            this.setState({
                walletFrom: this.props.wallets[0],
                walletTo: this.props.wallets[1]
            });
        }
    }
    
    render() {
        const { wallets, rates } = this.props;
        const { walletFrom, walletTo, amount } = this.state;
        if (!walletFrom && !walletTo) {
            return <span>Loading wallets...</span>;
        }
        if (!(rates && rates[walletFrom.currency])) {
            return <span>Loading rates...</span>;
        }

        const walletFromIndex = wallets.findIndex(w => w.id === walletFrom.id);
        const walletToIndex = wallets.findIndex(w => w.id === walletTo.id);

        const rateFrom = rates[walletFrom.currency][walletTo.currency];
        const rateTo = 1 / rateFrom;

        const amountTo = amount * rateFrom;

        return (
            <div>
                <div>
                    <button onClick={this.submitExchange}>Exchange</button>
                </div>
                <WalletPicker
                    type="from"
                    current={walletFrom}
                    currentIndex={walletFromIndex}
                    onWalletChange={this.onWalletChange}
                    onBallanceChange={this.onBallanceChange}
                    walletsAmount={wallets.length}
                    rate={rateFrom}
                    exchangeCurrency={walletTo.currency}
                    amount={amount}
                />
                <hr />
                <WalletPicker
                    type="to"
                    current={walletTo}
                    currentIndex={walletToIndex}
                    onBallanceChange={this.onBallanceChange}
                    onWalletChange={this.onWalletChange}
                    walletsAmount={wallets.length}
                    rate={rateTo}
                    exchangeCurrency={walletFrom.currency}
                    amount={amountTo}
                />
            </div>
        );
    }

    onBallanceChange(type: WalletType, value: string) {
        const rateFrom = this.props.rates[this.state.walletFrom.currency][this.state.walletTo.currency];
        const rateTo = 1 / rateFrom;
        type === 'from' && this.setState({
            amount: Number(value)
        });
        type === 'to' && this.setState({
            amount: Number(value) * rateTo
        });
    }

    onWalletChange(type: WalletType, index: number) {
        this.setState({
            walletFrom: type === 'from' ? this.props.wallets[index] : this.state.walletFrom,
            walletTo: type === 'to' ? this.props.wallets[index] : this.state.walletTo
        }, () => {
            type === 'from' && this.props.updateBaseCurrency(this.state.walletFrom.currency);
        });
    }

    submitExchange() {
        this.props.submitExchange({
            walletFromId: this.state!.walletFrom.id,
            walletToId: this.state!.walletTo.id,
            amount: this.state.amount,
            rate: this.props.rates[this.state.walletFrom.currency][this.state.walletTo.currency]
        });
    }
}

export const App = appConnector(AppPresenter);
