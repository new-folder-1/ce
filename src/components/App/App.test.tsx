import React from 'react';
import { shallow } from 'enzyme';

import { AppPresenter as App, AppProps } from './App';

const baseProps: AppProps = {
    rates: {
        'USD': {
            'EUR': 0.8992,
            'USD': 1,
            'GBP': 0.7654
        }
    },
    wallets: [
        {
            currency: 'USD',
            amount: 200,
            id: '1'
        },
        {
            currency: 'GBP',
            amount: 300,
            id: '2'
        },
        {
            currency: 'EUR',
            amount: 100.20,
            id: '3'
        },
    ],
    fetchWallets: jest.fn(),
    updateBaseCurrency: jest.fn(),
    submitExchange: jest.fn()
};

describe('App', () => {
    it('snapshot renders', () => {
        const component = shallow(<App
            {...baseProps}
        />);
        expect(component).toMatchSnapshot();
    });
});
