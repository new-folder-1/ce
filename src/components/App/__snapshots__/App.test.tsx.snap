// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App snapshot renders 1`] = `
<div
  className="App"
>
  <div
    className="App-Header"
  >
    <span />
    <Memo()
      disabled={true}
      onClick={[Function]}
      text="Exchange"
      theme="action"
    />
  </div>
  <div
    className="App-Wallet App-Wallet_from"
  >
    <Component
      amountDisabled={false}
      current={
        Object {
          "amount": 200,
          "currency": "USD",
          "id": "1",
        }
      }
      exchangeCurrency="GBP"
      next={
        Object {
          "amount": 300,
          "currency": "GBP",
          "id": "2",
        }
      }
      onAmountChange={[Function]}
      onWalletChange={[Function]}
      rate={0.765404}
      type="from"
    />
  </div>
  <div
    className="App-Wallet App-Wallet_to"
  >
    <Component
      amountDisabled={false}
      current={
        Object {
          "amount": 300,
          "currency": "GBP",
          "id": "2",
        }
      }
      exchangeCurrency="USD"
      next={
        Object {
          "amount": 100.2,
          "currency": "EUR",
          "id": "3",
        }
      }
      onAmountChange={[Function]}
      onWalletChange={[Function]}
      prev={
        Object {
          "amount": 200,
          "currency": "USD",
          "id": "1",
        }
      }
      rate={1.306499574081139}
      type="to"
    />
  </div>
</div>
`;
